src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -0,0 +1,236 @@
+#include "postgres.h"
+#include "c.h"
+
+#include "distributed/citus_nodefuncs.h"
+#include "distributed/citus_ruleutils.h"
+#include "distributed/deparse_shard_query.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/multi_router_planner.h"
+#include "lib/stringinfo.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/nodes.h"
+#include "nodes/parsenodes.h"
+#include "nodes/pg_list.h"
+#include "utils/lsyscache.h"
+#include "utils/rel.h"
+
+
+static void ConvertRteToSubqueryWithEmptyResult(RangeTblEntry *rte);
+
+
+/*
+ * RebuildQueryStrings deparses the job query for each task to
+ * include execution-time changes such as function evaluation.
+ */
+void
+RebuildQueryStrings(Query *originalQuery, List *taskList)
+{
+	ListCell *taskCell = NULL;
+	Oid relationId = ((RangeTblEntry *) linitial(originalQuery->rtable))->relid;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		StringInfo newQueryString = makeStringInfo();
+		Query *query = originalQuery;
+
+		if (task->insertSelectQuery)
+		{
+			/* for INSERT..SELECT, adjust shard names in SELECT part */
+			RangeTblEntry *copiedInsertRte = NULL;
+			RangeTblEntry *copiedSubqueryRte = NULL;
+			Query *copiedSubquery = NULL;
+			List *relationShardList = task->relationShardList;
+			ShardInterval *shardInterval = LoadShardInterval(task->anchorShardId);
+
+			query = copyObject(originalQuery);
+
+			copiedInsertRte = ExtractInsertRangeTableEntry(query);
+			copiedSubqueryRte = ExtractSelectRangeTableEntry(query);
+			copiedSubquery = copiedSubqueryRte->subquery;
+
+			AddShardIntervalRestrictionToSelect(copiedSubquery, shardInterval);
+			ReorderInsertSelectTargetLists(query, copiedInsertRte, copiedSubqueryRte);
+
+			/* setting an alias simplifies deparsing of RETURNING */
+			if (copiedInsertRte->alias == NULL)
+			{
+				Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
+				copiedInsertRte->alias = alias;
+			}
+
+			UpdateRelationToShardNames((Node *) copiedSubquery, relationShardList);
+		}
+
+		deparse_shard_query(query, relationId, task->anchorShardId,
+							newQueryString);
+
+		ereport(DEBUG4, (errmsg("query before rebuilding: %s",
+								task->queryString)));
+		ereport(DEBUG4, (errmsg("query after rebuilding:  %s",
+								newQueryString->data)));
+
+		task->queryString = newQueryString->data;
+	}
+}
+
+
+/*
+ * UpdateRelationToShardNames walks over the query tree and appends shard ids to
+ * relations. It uses unique identity value to establish connection between a
+ * shard and the range table entry. If the range table id is not given a
+ * identity, than the relation is not referenced from the query, no connection
+ * could be found between a shard and this relation. Therefore relation is replaced
+ * by set of NULL values so that the query would work at worker without any problems.
+ *
+ */
+bool
+UpdateRelationToShardNames(Node *node, List *relationShardList)
```

Comment:
Adapted from UpdateRelationNames multi_router_planner.c

I moved it into its own file because it's now being called both directly from the router planner as well as from the router executor via RebuildQueryStrings, it thus didn't seem very specific to the router planner anymore, but rather to deparsing shard queries in general.